Add Source §1 features

remo5000 opened the pull request:
This intention of this PR is to facilitate testing of the SICP textbook. The features listed are implemented under the week 3 library, as it is the most similar to the chapter 1 specifications.

- Arrow functions (using a new `ArrowClosure`)
- More symbols according to the §1 specifications
    - `parse_int`
    - `runtime` 
    - `error`
    - `math_PI`,
    - and `undefined`
- A `SourceError`, `noDeclareReservedError` that throws an error upon declaring a variable with a reserved keyword.

Latest specifications: [source_1.pdf](https://github.com/source-academy/source-academy2/files/1999208/source_1.pdf)

remo5000 committed:
Integrate es6 implementation from source 1-1

Refer to the `slang` repository, under commit `7c1eafd`.

Added in this commit:
- ES6 support
- Addition of `const` and `let` keywords
- Prevention of variable redeclaration (for now)

What doesn't work:
- Arrow functions

ning-y committed:
Add arrow functions

ning-y committed:
Update source 'versions' with §

ning-y committed:
Fix some typescript errors

Mirrors slang repo commit c1d9e00

ning-y committed:
Add parseInt as per specifications

Source's parseInt requires that both first and second positional
arguments, the inputString and radix be provided. This is in contrast to
the native implementation of parseInt where the second radix argument
can be omitted. As a result, a wrapper function `parse_int` in `misc`
was defined to raise errors for bad arguments.

ning-y committed:
Add symbol undefined

As per source chapter 1 specifications

remo5000 committed:
Move prompt to source-v1

remo5000 committed:
Add built-in function error

Added right after display due to their semantic similarity.

ning-y committed:
Remove unspecified built-ins from source chap 1

Since we had last mapped source chapter 1 to source week 3, we now need
to remove symbols defined for source week 3 that are not defined in the
specifications for source chapter 1

ning-y committed:
Add symbol math_PI

As per the specification for source chapter 1.

ning-y committed:
Add built-in function runtime

As per specifications for source chapter 1

ning-y committed:
Enable built-in function display for source chap1

As per specifications. Display was already defined for source week 4,
but we mapped source chapter 1 to source week 3. So, display was moved
from week 4 to week 3.

ning-y committed:
Reorder defineSymbol's according to specification

Reorder the sequence of defineSymbols according to their order in the
specifications for source chapter 1.

ning-y committed:
Add SourceError for declaring over reserved words

Note that some reserved words define by the source specifications such
as `true` will raise a SyntaxError first before reaching this new error.

ning-y committed:
Update string representations of built-ins

These are stored in the __SOURCE__ attribute and accessed in interop.ts

ning-y committed:
Fix name of built-in 'parseInt' -> 'parse_int'

The native ECMAScript function is named `parseInt`, but source
specifications provide `parse_int`

ning-y commented:
@evansb If the changes are okay, could you help us deploy the changes onto the NUS sadm server? Prof Henz had requested that we make the changes live so that the example programs for the new textbook can be tested, but we don't want to take the risk of messing with the prod servers.

evansb commented on frontend/src/toolchain/interpreter.ts:
```diff
@@ -64,15 +65,29 @@ const handleError = (context: Context, error: SourceError) => {
       context,
       new errors.VariableRedeclaration(context.runtime.nodes[0]!, name)
     )
+  } else if (kind === 'let' || kind === 'var') {
```
This should be a parser rule, not runtime

evansb commented:
I don't have access to the SADM server though. I can give you the tarball and instruction to deploy

evansb commented:
I thought we agreed to use CLI?

remo5000 committed:
Modify mutable variable declaration to SourceError

Changed it from a runtime error into a SourceError, under the `rules`
folder (like noDeclareReserved)

remo5000 committed:
Remove erroneous error declaration

Missed out on removing these declarations (that were used to make a runtime error)

remo5000 committed:
Fix defineVariable() calls

The signature of defineVariable() changed in commit 944438 , but its
calls were not changed, causing a typescript error.

ning-y commented:
@evansb Sorry, we just pushed an important bugfix with 249d2dc.
